ElectricCar.cpp

```diff
@@ -1,16 +1,45 @@
 #include "ElectricCar.hpp"
 #include <iostream>
 
-ElectricCar::ElectricCar(ElectricEngine* engine)
-    : engine_(engine)
-{
-    std::cout << __FUNCTION__ << std::endl;
+ElectricCar::ElectricCar(std::unique_ptr<ElectricEngine> engine)
+    : engine_(std::move(engine)) {
+  std::cout << __FUNCTION__ << std::endl;
 }
 
-ElectricCar::~ElectricCar()       { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+ElectricCar::~ElectricCar() { std::cout << __FUNCTION__ << std::endl; }
 
+void ElectricCar::charge() { std::cout << __FUNCTION__ << std::endl; }
+
+ElectricEngine *ElectricCar::getElectricEngineData() const {
+  if(engine_ != nullptr)
+  return engine_.get();
+  else throw std::runtime_error("You can't get to data, because engine doesn't exists\n");
+}
+
+std::unique_ptr<ElectricEngine> ElectricCar::pullOutElectricEngine() {
+    if(speed == 0){
+      std::cout << "Electric engine is pulled out\n";
+  return std::move(engine_);
+    }else throw std::runtime_error("This car is moving, you can't pull out engine\n");
+}
+
+void ElectricCar::restore(){
+    charge();
+}
+
+void ElectricCar::accelerate(int a) {
```

Comment:
very nice but i suggest you not to use variables like 'a, b, c'

Reply:
+1